Cannot start Metabase on Turkish language OS

**Describe the bug**
Since 0.34.0, it hasn't been possible to start Metabase on a Turkish language OS unless manually changing the Java property `user.language=en`

**To Reproduce**
1. Start Metabase `java -Duser.language=tr -jar metabase.jar` - fails with:
```
05-30 15:31:21 DEBUG plugins.classloader :: Using NEWLY CREATED classloader as shared context classloader: clojure.lang.DynamicClassLoader@71e4b308
05-30 15:31:21 INFO metabase.util :: Metatabanı yükleniyor ...
05-30 15:31:21 INFO metabase.util :: JVM için kullanılabilir bellek: 1,9 GB
Exception in thread "main" java.lang.AssertionError: Assert failed: Invalid TemporalField: nil
(instance? TemporalField field)
	at metabase.util.date_2.parse.builder$temporal_field.invokeStatic(builder.clj:96)
	at metabase.util.date_2.parse.builder$temporal_field.invoke(builder.clj:92)
	at metabase.util.date_2.parse.builder$value$fn__18342.invoke(builder.clj:109)
	at metabase.util.date_2.parse.builder$fn__18312.invokeStatic(builder.clj:24)
	at metabase.util.date_2.parse.builder$fn__18312.invoke(builder.clj:17)
	at metabase.util.date_2.parse.builder$fn__18297$G__18292__18304.invoke(builder.clj:14)
	at metabase.util.date_2.parse.builder$fn__18314.invokeStatic(builder.clj:29)
	at metabase.util.date_2.parse.builder$fn__18314.invoke(builder.clj:17)
	at metabase.util.date_2.parse.builder$fn__18297$G__18292__18304.invoke(builder.clj:14)
	at metabase.util.date_2.parse.builder$optional$reify__18324.apply_section(builder.clj:41)
	at metabase.util.date_2.parse.builder$fn__18314.invokeStatic(builder.clj:29)
	at metabase.util.date_2.parse.builder$fn__18314.invoke(builder.clj:17)
	at metabase.util.date_2.parse.builder$fn__18297$G__18292__18304.invoke(builder.clj:14)
	at metabase.util.date_2.parse.builder$formatter.invokeStatic(builder.clj:163)
	at metabase.util.date_2.parse.builder$formatter.doInvoke(builder.clj:144)
	at clojure.lang.RestFn.invoke(RestFn.java:482)
	at metabase.util.date_2.parse__init.load(Unknown Source)
	at metabase.util.date_2.parse__init.<clinit>(Unknown Source)
	at java.base/java.lang.Class.forName0(Native Method)
	at java.base/java.lang.Class.forName(Class.java:398)
	at clojure.lang.RT.classForName(RT.java:2211)
	at clojure.lang.RT.classForName(RT.java:2220)
	at clojure.lang.RT.loadClassForName(RT.java:2239)
	at clojure.lang.RT.load(RT.java:449)
	at clojure.lang.RT.load(RT.java:424)
	at clojure.core$load$fn__6839.invoke(core.clj:6126)
	at clojure.core$load.invokeStatic(core.clj:6125)
	at clojure.core$load.doInvoke(core.clj:6109)
	at clojure.lang.RestFn.invoke(RestFn.java:408)
	at clojure.core$load_one.invokeStatic(core.clj:5908)
	at clojure.core$load_one.invoke(core.clj:5903)
	at clojure.core$load_lib$fn__6780.invoke(core.clj:5948)
	at clojure.core$load_lib.invokeStatic(core.clj:5947)
	at clojure.core$load_lib.doInvoke(core.clj:5928)
	at clojure.lang.RestFn.applyTo(RestFn.java:142)
	at clojure.core$apply.invokeStatic(core.clj:667)
	at clojure.core$load_libs.invokeStatic(core.clj:5989)
	at clojure.core$load_libs.doInvoke(core.clj:5969)
	at clojure.lang.RestFn.applyTo(RestFn.java:137)
	at clojure.core$apply.invokeStatic(core.clj:667)
	at clojure.core$require.invokeStatic(core.clj:6007)
	at clojure.core$require.doInvoke(core.clj:6007)
	at clojure.lang.RestFn.invoke(RestFn.java:619)
	at metabase.util.date_2$loading__6721__auto____18243.invoke(date_2.clj:1)
	at metabase.util.date_2__init.load(Unknown Source)
	at metabase.util.date_2__init.<clinit>(Unknown Source)
	at java.base/java.lang.Class.forName0(Native Method)
	at java.base/java.lang.Class.forName(Class.java:398)
	at clojure.lang.RT.classForName(RT.java:2211)
	at clojure.lang.RT.classForName(RT.java:2220)
	at clojure.lang.RT.loadClassForName(RT.java:2239)
	at clojure.lang.RT.load(RT.java:449)
	at clojure.lang.RT.load(RT.java:424)
	at clojure.core$load$fn__6839.invoke(core.clj:6126)
	at clojure.core$load.invokeStatic(core.clj:6125)
	at clojure.core$load.doInvoke(core.clj:6109)
	at clojure.lang.RestFn.invoke(RestFn.java:408)
	at clojure.core$load_one.invokeStatic(core.clj:5908)
	at clojure.core$load_one.invoke(core.clj:5903)
	at clojure.core$load_lib$fn__6780.invoke(core.clj:5948)
	at clojure.core$load_lib.invokeStatic(core.clj:5947)
	at clojure.core$load_lib.doInvoke(core.clj:5928)
	at clojure.lang.RestFn.applyTo(RestFn.java:142)
	at clojure.core$apply.invokeStatic(core.clj:667)
	at clojure.core$load_libs.invokeStatic(core.clj:5985)
	at clojure.core$load_libs.doInvoke(core.clj:5969)
	at clojure.lang.RestFn.applyTo(RestFn.java:137)
	at clojure.core$apply.invokeStatic(core.clj:667)
	at clojure.core$require.invokeStatic(core.clj:6007)
	at clojure.core$require.doInvoke(core.clj:6007)
	at clojure.lang.RestFn.invoke(RestFn.java:512)
	at metabase.db.jdbc_protocols$loading__6721__auto____15348.invoke(jdbc_protocols.clj:1)
	at metabase.db.jdbc_protocols__init.load(Unknown Source)
	at metabase.db.jdbc_protocols__init.<clinit>(Unknown Source)
	at java.base/java.lang.Class.forName0(Native Method)
	at java.base/java.lang.Class.forName(Class.java:398)
	at clojure.lang.RT.classForName(RT.java:2211)
	at clojure.lang.RT.classForName(RT.java:2220)
	at clojure.lang.RT.loadClassForName(RT.java:2239)
	at clojure.lang.RT.load(RT.java:449)
	at clojure.lang.RT.load(RT.java:424)
	at clojure.core$load$fn__6839.invoke(core.clj:6126)
	at clojure.core$load.invokeStatic(core.clj:6125)
	at clojure.core$load.doInvoke(core.clj:6109)
	at clojure.lang.RestFn.invoke(RestFn.java:408)
	at clojure.core$load_one.invokeStatic(core.clj:5908)
	at clojure.core$load_one.invoke(core.clj:5903)
	at clojure.core$load_lib$fn__6780.invoke(core.clj:5948)
	at clojure.core$load_lib.invokeStatic(core.clj:5947)
	at clojure.core$load_lib.doInvoke(core.clj:5928)
	at clojure.lang.RestFn.applyTo(RestFn.java:142)
	at clojure.core$apply.invokeStatic(core.clj:667)
	at clojure.core$load_libs.invokeStatic(core.clj:5989)
	at clojure.core$load_libs.doInvoke(core.clj:5969)
	at clojure.lang.RestFn.applyTo(RestFn.java:137)
	at clojure.core$apply.invokeStatic(core.clj:667)
	at clojure.core$require.invokeStatic(core.clj:6007)
	at clojure.core$require.doInvoke(core.clj:6007)
	at clojure.lang.RestFn.invoke(RestFn.java:805)
	at metabase.db$loading__6721__auto____15300.invoke(db.clj:1)
	at metabase.db__init.load(Unknown Source)
	at metabase.db__init.<clinit>(Unknown Source)
	at java.base/java.lang.Class.forName0(Native Method)
	at java.base/java.lang.Class.forName(Class.java:398)
	at clojure.lang.RT.classForName(RT.java:2211)
	at clojure.lang.RT.classForName(RT.java:2220)
	at clojure.lang.RT.loadClassForName(RT.java:2239)
	at clojure.lang.RT.load(RT.java:449)
	at clojure.lang.RT.load(RT.java:424)
	at clojure.core$load$fn__6839.invoke(core.clj:6126)
	at clojure.core$load.invokeStatic(core.clj:6125)
	at clojure.core$load.doInvoke(core.clj:6109)
	at clojure.lang.RestFn.invoke(RestFn.java:408)
	at clojure.core$load_one.invokeStatic(core.clj:5908)
	at clojure.core$load_one.invoke(core.clj:5903)
	at clojure.core$load_lib$fn__6780.invoke(core.clj:5948)
	at clojure.core$load_lib.invokeStatic(core.clj:5947)
	at clojure.core$load_lib.doInvoke(core.clj:5928)
	at clojure.lang.RestFn.applyTo(RestFn.java:142)
	at clojure.core$apply.invokeStatic(core.clj:667)
	at clojure.core$load_libs.invokeStatic(core.clj:5989)
	at clojure.core$load_libs.doInvoke(core.clj:5969)
	at clojure.lang.RestFn.applyTo(RestFn.java:137)
	at clojure.core$apply.invokeStatic(core.clj:667)
	at clojure.core$require.invokeStatic(core.clj:6007)
	at clojure.core$require.doInvoke(core.clj:6007)
	at clojure.lang.RestFn.invoke(RestFn.java:619)
	at metabase.core$loading__6721__auto____71189.invoke(core.clj:2)
	at metabase.core__init.load(Unknown Source)
	at metabase.core__init.<clinit>(Unknown Source)
	at java.base/java.lang.Class.forName0(Native Method)
	at java.base/java.lang.Class.forName(Class.java:398)
	at clojure.lang.RT.classForName(RT.java:2211)
	at clojure.lang.RT.classForName(RT.java:2220)
	at clojure.lang.RT.loadClassForName(RT.java:2239)
	at clojure.lang.RT.load(RT.java:449)
	at clojure.lang.RT.load(RT.java:424)
	at clojure.core$load$fn__6839.invoke(core.clj:6126)
	at clojure.core$load.invokeStatic(core.clj:6125)
	at clojure.core$load.doInvoke(core.clj:6109)
	at clojure.lang.RestFn.invoke(RestFn.java:408)
	at clojure.lang.Var.invoke(Var.java:384)
	at clojure.lang.Util.loadWithClass(Util.java:251)
	at metabase.core.<clinit>(Unknown Source)
```

**Information about your Metabase Installation:**
Metabase 0.34.0 thru 0.35.4 on Linux and Windows

**Severity**
Setting P1, since we just announced Turkish language support again in 0.35.4, so this will likely become a common issue.

**Additional context**
https://discourse.metabase.com/t/metabase-install-error-on-windows-10/10323
Related to #9073

--- a/src/metabase/util/date_2/common.clj
+++ b/src/metabase/util/date_2/common.clj
@@ -1,5 +1,6 @@
 (ns metabase.util.date-2.common
-  (:require [clojure.string :as str])
+  (:require [clojure.string :as str]
+            [metabase.util :as u])
   (:import [java.time.temporal ChronoField IsoFields TemporalField WeekFields]))
 
 ;; TODO - not sure this belongs here, it seems to be a bit more general than just `date-2`.
@@ -12,7 +13,7 @@
   ([^Class klass ^Class target-class]
    (into {} (for [^java.lang.reflect.Field f (.getFields klass)
                   :when                      (.isAssignableFrom target-class (.getType f))]
-              [(keyword (str/lower-case (str/replace (.getName f) #"_" "-")))
+              [(keyword (u/lower-case-en (str/replace (.getName f) #"_" "-")))
                (.get f nil)]))))
 
 (def ^TemporalField temporal-field

--- a/src/metabase/util/date_2/parse/builder.clj
+++ b/src/metabase/util/date_2/parse/builder.clj
@@ -9,7 +9,8 @@
 
   TODO - this is a prime library candidate."
   (:require [metabase.util.date-2.common :as common])
-  (:import [java.time.format DateTimeFormatter DateTimeFormatterBuilder SignStyle] java.time.temporal.TemporalField))
+  (:import [java.time.format DateTimeFormatter DateTimeFormatterBuilder SignStyle]
+           java.time.temporal.TemporalField))
 
 (defprotocol ^:private Section
   (^:private apply-section [this builder]))

--- a/test/metabase/test.clj
+++ b/test/metabase/test.clj
@@ -162,6 +162,7 @@
   round-all-decimals
   scheduler-current-tasks
   throw-if-called
+  with-locale
   with-log-messages
   with-log-messages-for-level
   with-log-level

--- a/test/metabase/test/util.clj
+++ b/test/metabase/test/util.clj
@@ -29,6 +29,7 @@
             [toucan.db :as db]
             [toucan.util.test :as tt])
   (:import java.util.concurrent.TimeoutException
+           java.util.Locale
            org.apache.log4j.Logger
            [org.quartz CronTrigger JobDetail JobKey Scheduler Trigger]))
 
@@ -708,3 +709,28 @@
 
     :else
     x))
+
+(defmacro exception-and-message
+  "Invokes `body`, catches the exception and returns a map with the exception class, message and data"
+  [& body]
+  `(try
+     ~@body
+     (catch Exception e#
+       {:ex-class (class e#)
+        :msg      (.getMessage e#)
+        :data     (ex-data e#)})))
+
+(defn call-with-locale
+  "Sets the default locale temporarily to `locale-tag`, then invokes `f` and reverts the locale change"
+  [locale-tag f]
+  (let [current-locale (Locale/getDefault)]
+    (try
+      (Locale/setDefault (Locale/forLanguageTag locale-tag))
+      (f)
+      (finally
+        (Locale/setDefault current-locale)))))
+
+(defmacro with-locale
+  "Allows a test to override the locale temporarily"
+  [locale-tag & body]
+  `(call-with-locale ~locale-tag (fn [] ~@body)))

--- a/test/metabase/util/date_2_test.clj
+++ b/test/metabase/util/date_2_test.clj
@@ -3,8 +3,11 @@
              [string :as str]
              [test :refer :all]]
             [java-time :as t]
+            [metabase.test :as mt]
             [metabase.test.util.timezone :as tu.timezone]
-            [metabase.util.date-2 :as u.date]))
+            [metabase.util.date-2 :as u.date]
+            [metabase.util.date-2.common :as u.date.common])
+  (:import java.time.temporal.ChronoField))
 
 (deftest parse-test
   ;; system timezone should not affect the way strings are parsed
@@ -400,3 +403,8 @@
             (is (= false
                    (u.date/older-than? t (t/months 2)))
                 (format "%s did not happen before 2019-10-03" (pr-str t)))))))))
+
+(deftest static-instances-locale-test
+  (testing "in the Turkish locale, :minute-of-hour can be found"
+    (mt/with-locale "tr"
+      (is (some? (:minute-of-hour (u.date.common/static-instances ChronoField)))))))

--- a/test/metabase/util/honeysql_extensions_test.clj
+++ b/test/metabase/util/honeysql_extensions_test.clj
@@ -3,9 +3,9 @@
             [honeysql
              [core :as hsql]
              [format :as hformat]]
+            [metabase.test :as mt]
             [metabase.util.honeysql-extensions :as hx])
-  (:import java.util.Locale
-           metabase.util.honeysql_extensions.Identifier))
+  (:import metabase.util.honeysql_extensions.Identifier))
 
 (deftest format-test
   (testing "Basic format test not including a specific quoting option"
@@ -95,26 +95,13 @@
     (is (= (Identifier. :field ["keyword" "qualified/keyword"])
            (hx/identifier :field :keyword :qualified/keyword)))))
 
-(defn- call-with-locale
-  "Sets the default locale temporarily to `locale-tag`, then invokes `f` and reverts the locale change"
-  [locale-tag f]
-  (let [current-locale (Locale/getDefault)]
-    (try
-      (Locale/setDefault (Locale/forLanguageTag locale-tag))
-      (f)
-      (finally
-        (Locale/setDefault current-locale)))))
-
-(defmacro ^:private with-locale [locale-tag & body]
-  `(call-with-locale ~locale-tag (fn [] ~@body)))
-
 (deftest h2-quoting-test
   (testing (str "We provide our own quoting function for `:h2` databases. We quote and uppercase the identifier. Using "
                 "Java's toUpperCase method is surprisingly locale dependent. When uppercasing a string in a language "
                 "like Turkish, it can turn an i into an İ. This test converts a keyword with an `i` in it to verify "
                 "that we convert the identifier correctly using the english locale even when the user has changed the "
                 "locale to Turkish")
-    (with-locale "tr"
+    (mt/with-locale "tr"
       (is (= ["\"SETTING\""]
              (hformat/format :setting :quoting :h2))))))
 

